### PR TITLE
TS examples for Sign In confirmation

### PR DESCRIPTION
**Pull request by devnotfound**
#### Description of changes:
This PR adds TypeScript examples for the following page under the authentication category -
Sign up, Sign in & Sign out
This PR also fixes semantic issues with the associated JavaScript examples

### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [ ] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [ ] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [X] JS
- [ ] iOS
- [ ] Android
- [ ] Flutter
- [ ] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [X] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [ ] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [ ] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [ ] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://link.com)` 
            HTML: `<a href="https://link.com">link</a>`_

### When this PR is ready to merge, please check the box below
- [x] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: src/fragments/lib/auth/js/emailpassword.mdx [hunk 1]
```diff
@@ -288,9 +288,55 @@ Based on the two parameters returned in the user object, you can then call the c
...
+
+```ts
 import { Auth } from 'aws-amplify';
 
+type ConfirmSignInWithMFAParameters = {
+  username: string;
+  password: string;
+};
+
+type MFAType = 'SMS_MFA' | 'SOFTWARE_TOKEN_MFA' | null;
```
Comment (israx) [inline src/fragments/lib/auth/js/emailpassword.mdx:302]:
```suggestion
type MFAType = 'SMS_MFA' | 'SOFTWARE_TOKEN_MFA';
```

Comment (israx) [reply src/fragments/lib/auth/js/emailpassword.mdx:302]: `null` is not a valid `MFAType`

File: src/fragments/lib/auth/js/emailpassword.mdx [hunk 2]
```diff
@@ -288,9 +288,55 @@ Based on the two parameters returned in the user object, you can then call the c
...
 If the user has SMS Multi-Factor Authentication (MFA) enabled, Cognito responds with a challenge that requires the user to provide an SMS verification code. This is where you can trigger `Auth.confirmSignIn()` and pass in the code provided by the user.  Amplify will then verify with Cognito that the SMS code is valid and complete the sign in process via Cognito returning access, ID, and refresh tokens that are all handled by Amplify internally.
 
-```javascript
+<BlockSwitcher>
+<Block name="TypeScript">
+
+```ts
 import { Auth } from 'aws-amplify';
 
+type ConfirmSignInWithMFAParameters = {
```
Comment (israx) [inline src/fragments/lib/auth/js/emailpassword.mdx:297]: these `type` name should be generic to handle any `signIn` call. I don't think a project would have different `sign-in` calls all over the code base to handle different purposes.

Comment (israx) [inline src/fragments/lib/auth/js/emailpassword.mdx:297]:
```suggestion
type SignInParameters = {
```

File: src/fragments/lib/auth/js/emailpassword.mdx [hunk 3]
```diff
@@ -288,9 +288,55 @@ Based on the two parameters returned in the user object, you can then call the c
...
+  username: string;
+  password: string;
+};
+
+type MFAType = 'SMS_MFA' | 'SOFTWARE_TOKEN_MFA' | null;
+
+export async function signInWithMFA({ username, password }: ConfirmSignInWithMFAParameters) {
```
Comment (israx) [inline src/fragments/lib/auth/js/emailpassword.mdx:304]: applying above suggestion

Comment (israx) [inline src/fragments/lib/auth/js/emailpassword.mdx:304]: I think `signInWithMFA` should have a generic name. e.g `handleSignIn`

File: src/fragments/lib/auth/js/emailpassword.mdx [hunk 4]
```diff
@@ -319,64 +365,157 @@ async function signInWithMFA() {
...
 
 If you have implemented custom challenges within your authentication flow, Cognito will respond to a sign in attempt with a `challengeName` and `challengeParam`.  Once again, Cognito will validate the the user's response and provide the necessary tokens if successful and resolves the `Auth.signIn()` promise.
 
+<BlockSwitcher>
+<Block name="TypeScript">
+
+```ts
+import { Auth } from 'aws-amplify';
+
+type ConfirmSignInWithCustomChallenge = {
```
Comment (israx) [inline src/fragments/lib/auth/js/emailpassword.mdx:381]: ditto

File: src/fragments/lib/auth/js/emailpassword.mdx [hunk 5]
```diff
@@ -319,64 +365,157 @@ async function signInWithMFA() {
...
+  username: string;
+  password: string;
+};
+
+export async function signInWithCustomChallenge({
```
Comment (israx) [inline src/fragments/lib/auth/js/emailpassword.mdx:386]: ditto

File: src/fragments/lib/auth/js/emailpassword.mdx [hunk 6]
```diff
@@ -319,64 +365,157 @@ async function signInWithMFA() {
...
+  username,
+  password,
+}: ConfirmSignInWithCustomChallenge) {
```
Comment (israx) [inline src/fragments/lib/auth/js/emailpassword.mdx:389]: ditto

File: src/fragments/lib/auth/js/emailpassword.mdx [hunk 7]
```diff
@@ -319,64 +365,157 @@ async function signInWithMFA() {
...
+
+export async function signInWithNewPassword({ username, password }: ConfirmSignInWithNewPassword) {
+  try {
+    const user = await Auth.signIn(username, password);
+    if (user.challengeName === 'NEW_PASSWORD_REQUIRED') {
+      // Required attributes are OPTIONAL, but if they were configured
+      // you need to get these in addition to new password from UI inputs
+      const { requiredAttributes } = user.challengeParam;
+      // Example if email and phone_number were required attributes
+      const { inputEmail, inputPhoneNumber, newPassword } = await getInfoFromUserInput();
```
Comment (israx) [inline src/fragments/lib/auth/js/emailpassword.mdx:464]: attributes should conform to the standard OIDC  format `email`, `phone_number`

File: src/fragments/lib/auth/js/emailpassword.mdx [hunk 8]
```diff
@@ -319,64 +365,157 @@ async function signInWithMFA() {
...
+      const loggedUser = await Auth.completeNewPassword(
+        user, // the Cognito User Object
+        newPassword, // the new password
+        // OPTIONAL, the required attributes
+        {
+          inputEmail,
```
Comment (israx) [inline src/fragments/lib/auth/js/emailpassword.mdx:470]: applying above suggestion

File: src/fragments/lib/auth/js/emailpassword.mdx [hunk 9]
```diff
@@ -319,64 +365,157 @@ async function signInWithMFA() {
...
     const user = await Auth.signIn(username, password);
-    if (user.challengeName === 'NEW_PASSWORD_REQUIRED`) {
-      // Required attributes are OPTIONAL, but if they were configured 
+    if (user.challengeName === 'NEW_PASSWORD_REQUIRED') {
+      // Required attributes are OPTIONAL, but if they were configured
       // you need to get these in addition to new password from UI inputs
       const { requiredAttributes } = user.challengeParam;
       // Example if email and phone_number were required attributes
-      const { username, email, phone_number } = getInfoFromUserInput();
+      const { inputEmail, inputPhoneNumber, newPassword } = getInfoFromUserInput();
```
Comment (israx) [inline src/fragments/lib/auth/js/emailpassword.mdx:499]: ditto

File: src/fragments/lib/auth/js/emailpassword.mdx [hunk 10]
```diff
@@ -319,64 +365,157 @@ async function signInWithMFA() {
...
 
       const loggedUser = await Auth.completeNewPassword(
         user, // the Cognito User Object
         newPassword, // the new password
         // OPTIONAL, the required attributes
         {
-          email,
-          phone_number
+          inputEmail,
```
Comment (israx) [inline src/fragments/lib/auth/js/emailpassword.mdx:506]: ditto

File: src/fragments/lib/auth/js/emailpassword.mdx [hunk 11]
```diff
@@ -319,64 +365,157 @@ async function signInWithMFA() {
...
+          inputPhoneNumber,
```
Comment (israx) [inline src/fragments/lib/auth/js/emailpassword.mdx:507]: ditto

**Review by israx — COMMENTED** [review]
PR looks good. Just a few comments bellow

File: src/fragments/lib/auth/js/emailpassword.mdx (hunk 7 above)
Comment (cwomack) [inline src/fragments/lib/auth/js/emailpassword.mdx:464]:
```suggestion
      const { email, phone_number, newPassword } = await getInfoFromUserInput();
```

File: src/fragments/lib/auth/js/emailpassword.mdx (hunk 8 above)
Comment (cwomack) [inline src/fragments/lib/auth/js/emailpassword.mdx:470]:
```suggestion
          email,
```

File: src/fragments/lib/auth/js/emailpassword.mdx [hunk 12]
```diff
@@ -319,64 +365,157 @@ async function signInWithMFA() {
...
+      const { requiredAttributes } = user.challengeParam;
+      // Example if email and phone_number were required attributes
+      const { inputEmail, inputPhoneNumber, newPassword } = await getInfoFromUserInput();
+      const loggedUser = await Auth.completeNewPassword(
+        user, // the Cognito User Object
+        newPassword, // the new password
+        // OPTIONAL, the required attributes
+        {
+          inputEmail,
+          inputPhoneNumber,
```
Comment (cwomack) [inline src/fragments/lib/auth/js/emailpassword.mdx:471]:
```suggestion
          phone_number,
```

File: src/fragments/lib/auth/js/emailpassword.mdx (hunk 9 above)
Comment (cwomack) [inline src/fragments/lib/auth/js/emailpassword.mdx:499]:
```suggestion
      const { email, phone_number, newPassword } = getInfoFromUserInput();
```

File: src/fragments/lib/auth/js/emailpassword.mdx (hunk 10 above)
Comment (cwomack) [inline src/fragments/lib/auth/js/emailpassword.mdx:506]:
```suggestion
          email,
```

File: src/fragments/lib/auth/js/emailpassword.mdx (hunk 11 above)
Comment (cwomack) [inline src/fragments/lib/auth/js/emailpassword.mdx:507]:
```suggestion
          phone_number,
```

File: src/fragments/lib/auth/js/emailpassword.mdx [hunk 13]
```diff
@@ -319,64 +365,157 @@ async function signInWithMFA() {
...
+
+type ConfirmSignInWithNewPassword = {
+  username: string;
+  password: string;
+};
+
+type UserInput = {
+  inputUsername: string;
+  inputEmail: string;
+  inputPhoneNumber: string;
```
Comment (cwomack) [inline src/fragments/lib/auth/js/emailpassword.mdx:452]:
```suggestion
  username: string;
  email: string;
  phone_number: string;
```

Comment (cwomack) [reply src/fragments/lib/auth/js/emailpassword.mdx:452]: update to standard OIDC naming for attributes within the type

**Review by cwomack — APPROVED** [review]
Updated with standard OIDC attribute names

**Review by israx — APPROVED** [review]
LGTM!

**Review by abdallahshaban557 — APPROVED** [review]
LGTM

**Review by timngyn — APPROVED** [review]

**Pull request merged**